{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "import time\n",
    "\n",
    "import requests\n",
    "from bs4 import BeautifulSoup\n",
    "import pandas as pd\n",
    "from typing import List, Dict\n",
    "from datetime import datetime\n",
    "import os\n",
    "import hashlib\n",
    "import json\n",
    "from pathlib import Path\n",
    "import re\n",
    "\n",
    "from io import BytesIO\n",
    "from typing import Optional\n",
    "\n",
    "import requests\n",
    "from scipy import sparse\n",
    "from io import BytesIO\n",
    "from urllib import response\n",
    "\n",
    "import pathlib\n",
    "from multiprocessing import Pool\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "_malignant_cell = \"Malignant cells\"\n",
    "_cell_type = \"cell_type\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n",
    "content = requests.get(\"https://ngdc.cncb.ac.cn/cancerscem/downloads/CancerSCEM.metadata.xlsx\").content\n",
    "\n",
    "metadata = pd.read_excel(BytesIO(content))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Project ID</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th>Sample Type</th>\n",
       "      <th>Data source</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>GEO</th>\n",
       "      <th>SRA</th>\n",
       "      <th>SRS</th>\n",
       "      <th>SRX</th>\n",
       "      <th>SRR</th>\n",
       "      <th>...</th>\n",
       "      <th>Project abstract</th>\n",
       "      <th>Construction protocol</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Instrument</th>\n",
       "      <th>Strategy</th>\n",
       "      <th>Library Layout</th>\n",
       "      <th>Cell count</th>\n",
       "      <th>Publication</th>\n",
       "      <th>PMID</th>\n",
       "      <th>DOI</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MCC-001</td>\n",
       "      <td>MCC-001-01-1A</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA483959</td>\n",
       "      <td>GSE117988</td>\n",
       "      <td>SRP155988</td>\n",
       "      <td>SRS3693909</td>\n",
       "      <td>SRX4579480</td>\n",
       "      <td>SRR7722938</td>\n",
       "      <td>...</td>\n",
       "      <td>Serial blood draws from a single patient and s...</td>\n",
       "      <td>PBMC were isolated with standard FICOLL, tumor...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>5477</td>\n",
       "      <td>K.G. Paulson et al., Acquired cancer resistanc...</td>\n",
       "      <td>30250229</td>\n",
       "      <td>https://doi.org/10.1038/s41467-018-06300-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MCC-001</td>\n",
       "      <td>MCC-001-02-1A</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA483959</td>\n",
       "      <td>GSE117988</td>\n",
       "      <td>SRP155988</td>\n",
       "      <td>SRS3693908</td>\n",
       "      <td>SRX4579479</td>\n",
       "      <td>SRR7722937</td>\n",
       "      <td>...</td>\n",
       "      <td>Serial blood draws from a single patient and s...</td>\n",
       "      <td>PBMC were isolated with standard FICOLL, tumor...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>2401</td>\n",
       "      <td>K.G. Paulson et al., Acquired cancer resistanc...</td>\n",
       "      <td>30250229</td>\n",
       "      <td>https://doi.org/10.1038/s41467-018-06300-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MCC-001</td>\n",
       "      <td>MCC-001-03-1A</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA483959</td>\n",
       "      <td>GSE117988</td>\n",
       "      <td>SRP155988</td>\n",
       "      <td>SRS3693910</td>\n",
       "      <td>SRX4579481</td>\n",
       "      <td>SRR7722939</td>\n",
       "      <td>...</td>\n",
       "      <td>Serial blood draws from a single patient and s...</td>\n",
       "      <td>PBMC were isolated with standard FICOLL, tumor...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>1991</td>\n",
       "      <td>K.G. Paulson et al., Acquired cancer resistanc...</td>\n",
       "      <td>30250229</td>\n",
       "      <td>https://doi.org/10.1038/s41467-018-06300-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MCC-001</td>\n",
       "      <td>MCC-001-04-1A</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA483959</td>\n",
       "      <td>GSE117988</td>\n",
       "      <td>SRP155988</td>\n",
       "      <td>SRS3693911</td>\n",
       "      <td>SRX4579482</td>\n",
       "      <td>SRR7722940\\t</td>\n",
       "      <td>...</td>\n",
       "      <td>Serial blood draws from a single patient and s...</td>\n",
       "      <td>PBMC were isolated with standard FICOLL, tumor...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>1064</td>\n",
       "      <td>K.G. Paulson et al., Acquired cancer resistanc...</td>\n",
       "      <td>30250229</td>\n",
       "      <td>https://doi.org/10.1038/s41467-018-06300-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MCC-001</td>\n",
       "      <td>MCC-001-05-1A</td>\n",
       "      <td>PBMC</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA483959</td>\n",
       "      <td>GSE117988</td>\n",
       "      <td>SRP155988</td>\n",
       "      <td>SRS3693912</td>\n",
       "      <td>SRX4579483</td>\n",
       "      <td>SRR7722941</td>\n",
       "      <td>...</td>\n",
       "      <td>Serial blood draws from a single patient and s...</td>\n",
       "      <td>PBMC were isolated with standard FICOLL, tumor...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>Illumina HiSeq 2500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>4524</td>\n",
       "      <td>K.G. Paulson et al., Acquired cancer resistanc...</td>\n",
       "      <td>30250229</td>\n",
       "      <td>https://doi.org/10.1038/s41467-018-06300-3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1461</th>\n",
       "      <td>BRCA-134</td>\n",
       "      <td>BRCA-134-01-1A</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA983305</td>\n",
       "      <td>GSE234832</td>\n",
       "      <td>SRP443229</td>\n",
       "      <td>SRS17968241</td>\n",
       "      <td>SRX20672578</td>\n",
       "      <td>SRR24911501; SRR24911502</td>\n",
       "      <td>...</td>\n",
       "      <td>Our analysis identifies intratumoral component...</td>\n",
       "      <td>Cellular suspensions were loaded on a Chromium...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>328</td>\n",
       "      <td>Q. Song et al., Single-cell sequencing reveals...</td>\n",
       "      <td>37479733</td>\n",
       "      <td>https://doi.org/10.1038/s42003-023-05124-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1462</th>\n",
       "      <td>BRCA-134</td>\n",
       "      <td>BRCA-134-02-1A</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA983305</td>\n",
       "      <td>GSE234832</td>\n",
       "      <td>SRP443229</td>\n",
       "      <td>SRS17968242</td>\n",
       "      <td>SRX20672577</td>\n",
       "      <td>SRR24911503; SRR24911504</td>\n",
       "      <td>...</td>\n",
       "      <td>Our analysis identifies intratumoral component...</td>\n",
       "      <td>Cellular suspensions were loaded on a Chromium...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>654</td>\n",
       "      <td>Q. Song et al., Single-cell sequencing reveals...</td>\n",
       "      <td>37479733</td>\n",
       "      <td>https://doi.org/10.1038/s42003-023-05124-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1463</th>\n",
       "      <td>BRCA-134</td>\n",
       "      <td>BRCA-134-03-1A</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA983305</td>\n",
       "      <td>GSE234832</td>\n",
       "      <td>SRP443229</td>\n",
       "      <td>SRS17968240</td>\n",
       "      <td>SRX20672576</td>\n",
       "      <td>SRR24911505; SRR24911506</td>\n",
       "      <td>...</td>\n",
       "      <td>Our analysis identifies intratumoral component...</td>\n",
       "      <td>Cellular suspensions were loaded on a Chromium...</td>\n",
       "      <td>10X Genomics</td>\n",
       "      <td>NextSeq 500</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>743</td>\n",
       "      <td>Q. Song et al., Single-cell sequencing reveals...</td>\n",
       "      <td>37479733</td>\n",
       "      <td>https://doi.org/10.1038/s42003-023-05124-2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1464</th>\n",
       "      <td>BRCA-135</td>\n",
       "      <td>BRCA-135-01-1H</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA780425</td>\n",
       "      <td>GSE188807</td>\n",
       "      <td>SRP346107</td>\n",
       "      <td>SRS11071870</td>\n",
       "      <td>SRX13138681</td>\n",
       "      <td>SRR16946746</td>\n",
       "      <td>...</td>\n",
       "      <td>To obtain more information about the lymph nod...</td>\n",
       "      <td>The concentration of single-cell suspension wa...</td>\n",
       "      <td>GEXSCOPE</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>2239</td>\n",
       "      <td>K. Xu et al., Microenvironment components and ...</td>\n",
       "      <td>36148946</td>\n",
       "      <td>https://doi.org/10.3724/abbs.2022131</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1465</th>\n",
       "      <td>BRCA-135</td>\n",
       "      <td>BRCA-135-02-1H</td>\n",
       "      <td>Tumour</td>\n",
       "      <td>NCBI (GEO)</td>\n",
       "      <td>PRJNA780425</td>\n",
       "      <td>GSE188807</td>\n",
       "      <td>SRP346107</td>\n",
       "      <td>SRS11071871</td>\n",
       "      <td>SRX13138682</td>\n",
       "      <td>SRR16946747</td>\n",
       "      <td>...</td>\n",
       "      <td>To obtain more information about the lymph nod...</td>\n",
       "      <td>The concentration of single-cell suspension wa...</td>\n",
       "      <td>GEXSCOPE</td>\n",
       "      <td>Illumina NovaSeq 6000</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>PAIRED</td>\n",
       "      <td>4242</td>\n",
       "      <td>K. Xu et al., Microenvironment components and ...</td>\n",
       "      <td>36148946</td>\n",
       "      <td>https://doi.org/10.3724/abbs.2022131</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1466 rows × 34 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     Project ID       Sample ID Sample Type Data source   BioProject  \\\n",
       "0       MCC-001   MCC-001-01-1A      Tumour  NCBI (GEO)  PRJNA483959   \n",
       "1       MCC-001   MCC-001-02-1A      Tumour  NCBI (GEO)  PRJNA483959   \n",
       "2       MCC-001   MCC-001-03-1A        PBMC  NCBI (GEO)  PRJNA483959   \n",
       "3       MCC-001   MCC-001-04-1A        PBMC  NCBI (GEO)  PRJNA483959   \n",
       "4       MCC-001   MCC-001-05-1A        PBMC  NCBI (GEO)  PRJNA483959   \n",
       "...         ...             ...         ...         ...          ...   \n",
       "1461   BRCA-134  BRCA-134-01-1A      Tumour  NCBI (GEO)  PRJNA983305   \n",
       "1462   BRCA-134  BRCA-134-02-1A      Tumour  NCBI (GEO)  PRJNA983305   \n",
       "1463   BRCA-134  BRCA-134-03-1A      Tumour  NCBI (GEO)  PRJNA983305   \n",
       "1464   BRCA-135  BRCA-135-01-1H      Tumour  NCBI (GEO)  PRJNA780425   \n",
       "1465   BRCA-135  BRCA-135-02-1H      Tumour  NCBI (GEO)  PRJNA780425   \n",
       "\n",
       "            GEO        SRA          SRS          SRX  \\\n",
       "0     GSE117988  SRP155988   SRS3693909   SRX4579480   \n",
       "1     GSE117988  SRP155988   SRS3693908   SRX4579479   \n",
       "2     GSE117988  SRP155988   SRS3693910   SRX4579481   \n",
       "3     GSE117988  SRP155988   SRS3693911   SRX4579482   \n",
       "4     GSE117988  SRP155988   SRS3693912   SRX4579483   \n",
       "...         ...        ...          ...          ...   \n",
       "1461  GSE234832  SRP443229  SRS17968241  SRX20672578   \n",
       "1462  GSE234832  SRP443229  SRS17968242  SRX20672577   \n",
       "1463  GSE234832  SRP443229  SRS17968240  SRX20672576   \n",
       "1464  GSE188807  SRP346107  SRS11071870  SRX13138681   \n",
       "1465  GSE188807  SRP346107  SRS11071871  SRX13138682   \n",
       "\n",
       "                           SRR  ...  \\\n",
       "0                   SRR7722938  ...   \n",
       "1                   SRR7722937  ...   \n",
       "2                   SRR7722939  ...   \n",
       "3                 SRR7722940\\t  ...   \n",
       "4                   SRR7722941  ...   \n",
       "...                        ...  ...   \n",
       "1461  SRR24911501; SRR24911502  ...   \n",
       "1462  SRR24911503; SRR24911504  ...   \n",
       "1463  SRR24911505; SRR24911506  ...   \n",
       "1464               SRR16946746  ...   \n",
       "1465               SRR16946747  ...   \n",
       "\n",
       "                                       Project abstract  \\\n",
       "0     Serial blood draws from a single patient and s...   \n",
       "1     Serial blood draws from a single patient and s...   \n",
       "2     Serial blood draws from a single patient and s...   \n",
       "3     Serial blood draws from a single patient and s...   \n",
       "4     Serial blood draws from a single patient and s...   \n",
       "...                                                 ...   \n",
       "1461  Our analysis identifies intratumoral component...   \n",
       "1462  Our analysis identifies intratumoral component...   \n",
       "1463  Our analysis identifies intratumoral component...   \n",
       "1464  To obtain more information about the lymph nod...   \n",
       "1465  To obtain more information about the lymph nod...   \n",
       "\n",
       "                                  Construction protocol      Protocol  \\\n",
       "0     PBMC were isolated with standard FICOLL, tumor...  10X Genomics   \n",
       "1     PBMC were isolated with standard FICOLL, tumor...  10X Genomics   \n",
       "2     PBMC were isolated with standard FICOLL, tumor...  10X Genomics   \n",
       "3     PBMC were isolated with standard FICOLL, tumor...  10X Genomics   \n",
       "4     PBMC were isolated with standard FICOLL, tumor...  10X Genomics   \n",
       "...                                                 ...           ...   \n",
       "1461  Cellular suspensions were loaded on a Chromium...  10X Genomics   \n",
       "1462  Cellular suspensions were loaded on a Chromium...  10X Genomics   \n",
       "1463  Cellular suspensions were loaded on a Chromium...  10X Genomics   \n",
       "1464  The concentration of single-cell suspension wa...      GEXSCOPE   \n",
       "1465  The concentration of single-cell suspension wa...      GEXSCOPE   \n",
       "\n",
       "                 Instrument Strategy Library Layout Cell count  \\\n",
       "0       Illumina HiSeq 2500  RNA-Seq         PAIRED       5477   \n",
       "1       Illumina HiSeq 2500  RNA-Seq         PAIRED       2401   \n",
       "2       Illumina HiSeq 2500  RNA-Seq         PAIRED       1991   \n",
       "3       Illumina HiSeq 2500  RNA-Seq         PAIRED       1064   \n",
       "4       Illumina HiSeq 2500  RNA-Seq         PAIRED       4524   \n",
       "...                     ...      ...            ...        ...   \n",
       "1461            NextSeq 500  RNA-Seq         PAIRED        328   \n",
       "1462            NextSeq 500  RNA-Seq         PAIRED        654   \n",
       "1463            NextSeq 500  RNA-Seq         PAIRED        743   \n",
       "1464  Illumina NovaSeq 6000  RNA-Seq         PAIRED       2239   \n",
       "1465  Illumina NovaSeq 6000  RNA-Seq         PAIRED       4242   \n",
       "\n",
       "                                            Publication      PMID  \\\n",
       "0     K.G. Paulson et al., Acquired cancer resistanc...  30250229   \n",
       "1     K.G. Paulson et al., Acquired cancer resistanc...  30250229   \n",
       "2     K.G. Paulson et al., Acquired cancer resistanc...  30250229   \n",
       "3     K.G. Paulson et al., Acquired cancer resistanc...  30250229   \n",
       "4     K.G. Paulson et al., Acquired cancer resistanc...  30250229   \n",
       "...                                                 ...       ...   \n",
       "1461  Q. Song et al., Single-cell sequencing reveals...  37479733   \n",
       "1462  Q. Song et al., Single-cell sequencing reveals...  37479733   \n",
       "1463  Q. Song et al., Single-cell sequencing reveals...  37479733   \n",
       "1464  K. Xu et al., Microenvironment components and ...  36148946   \n",
       "1465  K. Xu et al., Microenvironment components and ...  36148946   \n",
       "\n",
       "                                             DOI  \n",
       "0     https://doi.org/10.1038/s41467-018-06300-3  \n",
       "1     https://doi.org/10.1038/s41467-018-06300-3  \n",
       "2     https://doi.org/10.1038/s41467-018-06300-3  \n",
       "3     https://doi.org/10.1038/s41467-018-06300-3  \n",
       "4     https://doi.org/10.1038/s41467-018-06300-3  \n",
       "...                                          ...  \n",
       "1461  https://doi.org/10.1038/s42003-023-05124-2  \n",
       "1462  https://doi.org/10.1038/s42003-023-05124-2  \n",
       "1463  https://doi.org/10.1038/s42003-023-05124-2  \n",
       "1464        https://doi.org/10.3724/abbs.2022131  \n",
       "1465        https://doi.org/10.3724/abbs.2022131  \n",
       "\n",
       "[1466 rows x 34 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import BytesIO\n",
    "from typing import Optional\n",
    "\n",
    "import requests\n",
    "from scipy import sparse\n",
    "\n",
    "\n",
    "def get_dataframe(url: str, cache_dir: str = \".cache\", compression: Optional[str] = None) -> Optional[pd.DataFrame]:\n",
    "    # Create cache directory if it doesn't exist\n",
    "    cache_path = Path(cache_dir)\n",
    "    cache_path.mkdir(exist_ok=True)\n",
    "    \n",
    "    # Create a filename from URL using hash to avoid invalid characters\n",
    "    url_hash = hashlib.md5(url.encode()).hexdigest()\n",
    "    cache_file = cache_path / f\"{url_hash}.parquet\"\n",
    "    \n",
    "    # Check if file exists in cache\n",
    "    if cache_file.exists():\n",
    "        return pd.read_parquet(cache_file)\n",
    "    \n",
    "    # If not in cache, download and store\n",
    "    response = requests.get(url, stream=True)\n",
    "    if response.status_code == 200:\n",
    "        content = BytesIO(response.content)\n",
    "        df = pd.read_csv(content, compression=compression, sep='\\t', index_col=0)\n",
    "        \n",
    "        # Save to cache\n",
    "        df.to_parquet(cache_file)\n",
    "        return df\n",
    "    \n",
    "    return None\n",
    "\n",
    "\n",
    "def get_anndata_from_sample_id(sample_id: str) -> Optional[sc.AnnData]:\n",
    "    count_url = f'https://ngdc.cncb.ac.cn/cancerscem/downloads/expMatrixes/{sample_id}.counts.matrix.tsv.gz'\n",
    "    celltype_url = f'https://ngdc.cncb.ac.cn/cancerscem/downloads/cell_annotation/{sample_id}.cell.type.txt'\n",
    "\n",
    "    cell_type = get_dataframe(celltype_url)\n",
    "\n",
    "    cell_type_counts = cell_type.value_counts()\n",
    "\n",
    "    if not _malignant_cell in cell_type_counts.index:\n",
    "        return \n",
    "\n",
    "    if cell_type_counts[_malignant_cell] < 50:\n",
    "        return\n",
    "\n",
    "    count_data = get_dataframe(count_url, compression=\"gzip\")\n",
    "\n",
    "    if (count_data is None) or (cell_type is None):\n",
    "        return\n",
    "\n",
    "    adata = sc.AnnData(count_data.transpose())\n",
    "    adata.X = sparse.csr_matrix(adata.X)\n",
    "    \n",
    "    if sample_id.startswith(\"CRC-016-\"):\n",
    "        adata.obs_names = adata.obs_names.str.replace(\".\", \"-\")\n",
    "        cell_type.index = cell_type.index.str.replace(\".\", \"-\")\n",
    "    \n",
    "    adata.obs = cell_type.loc[adata.obs_names].copy()\n",
    "    adata.obs[\"sample_id\"] = sample_id\n",
    "    return adata\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "  0%|          | 0/138 [00:00<?, ?it/s]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  1%|          | 1/138 [00:01<04:15,  1.87s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  1%|▏         | 2/138 [00:04<05:18,  2.34s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  2%|▏         | 3/138 [00:09<08:04,  3.59s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  3%|▎         | 4/138 [00:12<07:39,  3.43s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  4%|▎         | 5/138 [00:13<05:27,  2.46s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  4%|▍         | 6/138 [00:16<05:28,  2.49s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  6%|▌         | 8/138 [00:22<05:28,  2.53s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  7%|▋         | 9/138 [00:25<06:07,  2.85s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      "  8%|▊         | 11/138 [00:29<04:50,  2.28s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 10%|█         | 14/138 [00:35<03:59,  1.93s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 11%|█         | 15/138 [00:36<03:26,  1.68s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 12%|█▏        | 16/138 [00:42<05:38,  2.78s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 13%|█▎        | 18/138 [00:43<03:11,  1.60s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 14%|█▍        | 19/138 [01:01<12:56,  6.52s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 14%|█▍        | 20/138 [01:05<11:30,  5.85s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 17%|█▋        | 23/138 [03:16<1:18:44, 41.08s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 19%|█▉        | 26/138 [04:12<48:51, 26.17s/it]  /local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 22%|██▏       | 31/138 [08:23<1:12:30, 40.66s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 23%|██▎       | 32/138 [09:34<1:28:01, 49.83s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 24%|██▍       | 33/138 [13:46<3:13:20, 110.48s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 25%|██▍       | 34/138 [15:06<2:55:37, 101.32s/it]/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 27%|██▋       | 37/138 [18:58<2:03:14, 73.21s/it] /local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/anndata/_core/anndata.py:1754: UserWarning: Observation names are not unique. To make them unique, call `.obs_names_make_unique`.\n",
      "  utils.warn_names_duplicates(\"obs\")\n",
      " 28%|██▊       | 38/138 [23:36<1:02:07, 37.27s/it]\n"
     ]
    },
    {
     "ename": "OSError",
     "evalue": "[Errno 28] Error writing bytes to file. Detail: [errno 28] No space left on device",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRemoteTraceback\u001b[0m                           Traceback (most recent call last)",
      "\u001b[0;31mRemoteTraceback\u001b[0m: \n\"\"\"\nTraceback (most recent call last):\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/multiprocessing/pool.py\", line 125, in worker\n    result = (True, func(*args, **kwds))\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/multiprocessing/pool.py\", line 48, in mapstar\n    return list(map(*args))\n  File \"/tmp/ipykernel_18516/2479071727.py\", line 48, in get_anndata_from_sample_id\n    count_data = get_dataframe(count_url, compression=\"gzip\")\n  File \"/tmp/ipykernel_18516/2479071727.py\", line 28, in get_dataframe\n    df.to_parquet(cache_file)\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/pandas/util/_decorators.py\", line 333, in wrapper\n    return func(*args, **kwargs)\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/pandas/core/frame.py\", line 3113, in to_parquet\n    return to_parquet(\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/pandas/io/parquet.py\", line 480, in to_parquet\n    impl.write(\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/pandas/io/parquet.py\", line 228, in write\n    self.api.parquet.write_table(\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/pyarrow/parquet/core.py\", line 3127, in write_table\n    writer.write_table(table, row_group_size=row_group_size)\n  File \"/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/site-packages/pyarrow/parquet/core.py\", line 1108, in write_table\n    self.writer.write_table(table, row_group_size=row_group_size)\n  File \"pyarrow/_parquet.pyx\", line 1892, in pyarrow._parquet.ParquetWriter.write_table\n  File \"pyarrow/error.pxi\", line 91, in pyarrow.lib.check_status\nOSError: [Errno 28] Error writing bytes to file. Detail: [errno 28] No space left on device\n\"\"\"",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mOSError\u001b[0m                                   Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[32], line 8\u001b[0m\n\u001b[1;32m      5\u001b[0m sample_ids \u001b[38;5;241m=\u001b[39m  metadata[metadata[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mProject ID\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m==\u001b[39mproject_id][\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mSample ID\u001b[39m\u001b[38;5;124m\"\u001b[39m]\u001b[38;5;241m.\u001b[39mto_list()\n\u001b[1;32m      7\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m Pool(processes\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m10\u001b[39m) \u001b[38;5;28;01mas\u001b[39;00m pool:\n\u001b[0;32m----> 8\u001b[0m     results \u001b[38;5;241m=\u001b[39m \u001b[43mpool\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmap\u001b[49m\u001b[43m(\u001b[49m\u001b[43mget_anndata_from_sample_id\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msample_ids\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     10\u001b[0m results \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mlist\u001b[39m(\u001b[38;5;28mfilter\u001b[39m(\u001b[38;5;28;01mlambda\u001b[39;00m x: x, results))\n\u001b[1;32m     11\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m results: \n",
      "File \u001b[0;32m/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/multiprocessing/pool.py:364\u001b[0m, in \u001b[0;36mPool.map\u001b[0;34m(self, func, iterable, chunksize)\u001b[0m\n\u001b[1;32m    359\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mmap\u001b[39m(\u001b[38;5;28mself\u001b[39m, func, iterable, chunksize\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    360\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124;03m'''\u001b[39;00m\n\u001b[1;32m    361\u001b[0m \u001b[38;5;124;03m    Apply `func` to each element in `iterable`, collecting the results\u001b[39;00m\n\u001b[1;32m    362\u001b[0m \u001b[38;5;124;03m    in a list that is returned.\u001b[39;00m\n\u001b[1;32m    363\u001b[0m \u001b[38;5;124;03m    '''\u001b[39;00m\n\u001b[0;32m--> 364\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_map_async\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfunc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43miterable\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mmapstar\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mchunksize\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/local/home/fbarkmann/envs/cancerfoundation/lib/python3.9/multiprocessing/pool.py:771\u001b[0m, in \u001b[0;36mApplyResult.get\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    769\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_value\n\u001b[1;32m    770\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[0;32m--> 771\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_value\n",
      "\u001b[0;31mOSError\u001b[0m: [Errno 28] Error writing bytes to file. Detail: [errno 28] No space left on device"
     ]
    }
   ],
   "source": [
    "\n",
    "\n",
    "results_dir = pathlib.Path(\"data/CancerSCEM/\")\n",
    "results_dir.mkdir(exist_ok=True)\n",
    "for project_id in tqdm(metadata[\"Project ID\"].unique()):\n",
    "    \n",
    "    sample_ids =  metadata[metadata[\"Project ID\"]==project_id][\"Sample ID\"].to_list()\n",
    "    \n",
    "    with Pool(processes=10) as pool:\n",
    "        results = pool.map(get_anndata_from_sample_id, sample_ids)\n",
    "    \n",
    "    results = list(filter(lambda x: x, results))\n",
    "    if not results: \n",
    "        continue\n",
    "    adata = sc.concat(results)\n",
    "    adata.obs_names = adata.obs_names + \"_\" +adata.obs[\"sample_id\"].astype(str)\n",
    "\n",
    "    adata.write_h5ad(results_dir.joinpath(f\"{project_id.replace('-', '_')}.h5ad\"))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results is None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CRC-016-01-1A',\n",
       " 'CRC-016-02-1A',\n",
       " 'CRC-016-03-1A',\n",
       " 'CRC-016-04-1A',\n",
       " 'CRC-016-05-1A',\n",
       " 'CRC-016-06-1A',\n",
       " 'CRC-016-07-1A',\n",
       " 'CRC-016-08-1A',\n",
       " 'CRC-016-09-1A',\n",
       " 'CRC-016-10-1A',\n",
       " 'CRC-016-11-1A',\n",
       " 'CRC-016-12-1A',\n",
       " 'CRC-016-13-1A',\n",
       " 'CRC-016-14-1A',\n",
       " 'CRC-016-15-1A',\n",
       " 'CRC-016-16-1A',\n",
       " 'CRC-016-17-1A',\n",
       " 'CRC-016-18-1A',\n",
       " 'CRC-016-19-1A',\n",
       " 'CRC-016-20-1A',\n",
       " 'CRC-016-21-1A',\n",
       " 'CRC-016-22-1A',\n",
       " 'CRC-016-23-1A',\n",
       " 'CRC-016-24-1A',\n",
       " 'CRC-016-25-1A',\n",
       " 'CRC-016-26-1A',\n",
       " 'CRC-016-27-1A']"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sample_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cancerfoundation",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
